### PR TITLE
Feature/async census creation

### PR DESCRIPTION
**Pull request by lucasmenendez**
now when the user tries to create a new census, it is included into the database and generated in background, a new census column has been created on the database called 'published' that indicates if the current census is built and published on IP

### CODE REVIEW

**Comment by lucasmenendez** [comment]
New flow to create and publish a census: 

1. Create a token
```sh
wget --method POST \
  --header 'Accept: */*' \
  --header 'Content-Type: application/json' \
  --body-data '{"id": "0xa117000000f279D81A1D3cc75430fAA017FA5A2e","type": "erc20"}' \
  --output-document \
  - https://census3.dev.vocdoni.net/api/token
```
*Response*:
```
Ok
```

2. Create census with `strategyID` that includes the created token:
```sh
wget --method POST \
  --header 'Accept: */*' \
  --header 'Content-Type: application/json' \
  --body-data '{"strategyId": 1}' \
  --output-document \
  - https://census3.dev.vocdoni.net/api/census
```
*Response*:
```json
{
  "queueId": "7c208c69d85dc187725395124514baea3bc5498c"
}
```

3. Get the census creation status from the queue with the `queueID` resulting of previous request:
```sh
wget --method GET \
  --header 'Accept: */*' \
  --output-document \
  - https://census3.dev.vocdoni.net/api/census/queue/7c208c69d85dc187725395124514baea3bc5498c
```
*Response*:
```json
{
  "census": {
    "censusId": 1,
    "strategyId": 1,
    "merkleRoot": "8f2d3b5c15445e6ec3bfe6a5f52c2be1de916ca296ce7474982b4a66b34d3f30",
    "uri": "ipfs://bafybeihb5krnn2muubms27gn6egevrjxeb5z7ttp35eqzi3b33ejqwntka",
    "size": 10,
    "weight": "320822718213393032631430073323170650669553037360",
    "chainId": 5,
    "anonymous": true
  },
  "done": true,
  "error": null
}
```

File: api/censuses.go [hunk 1]
```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
...
 	req := &CreateCensusResquest{}
 	if err := json.Unmarshal(msg.Data, req); err != nil {
 		return ErrMalformedStrategyID
 	}
-	// get tokens associated to the strategy
-	internalCtx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
-	defer cancel()
+	// create and publish census merkle tree in background
+	queueID := capi.queue.Enqueue()
+	go func(req *CreateCensusResquest) {
```
Comment (p4u) [inline api/censuses.go:108]: You don't need to pass req as parameter. The goroutine will inherit the vars from the caller routine.

File: api/censuses.go [hunk 2]
```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
...
+		censusID, err := capi.createAndPublishCensus(req, queueID)
```
Comment (p4u) [inline api/censuses.go:109]: I think I'd do this part of the goroutine sync, also the ErrCensusAlreadyExists error, and leave only the queue.Update() ASYNC. 

Sync is always easier and safer :)

File: api/censuses.go [hunk 3]
```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
...
-		holders, err := qtx.TokenHoldersByTokenID(internalCtx, token.ID)
+		holders, err := qtx.TokenHoldersByTokenID(bgCtx, token.ID)
 		if err != nil {
 			if errors.Is(sql.ErrNoRows, err) {
 				continue
 			}
-			log.Errorw(err, "error getting token holders")
-			return ErrCantGetTokenHolders.Withf("for the token with address %s",
-				common.BytesToAddress(token.ID))
+			log.Errorf("error getting token holders of %s: %v", common.BytesToAddress(token.ID), err)
```
Comment (p4u) [inline api/censuses.go:191]: Don't print the error if you return it

File: api/censuses.go [hunk 4]
```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
...
+	if len(strategyTokens) == 0 {
+		log.Errorf("no tokens for strategy %d", req.StrategyID)
+		return -1, ErrNoStrategyTokens.With("no tokens for strategy")
+	}
+
+	// get the maximun current census ID to calculate the next one, if any
+	// census has been created yet, continue
+	lastCensusID, err := qtx.LastCensusID(bgCtx)
+	if err != nil && !errors.Is(sql.ErrNoRows, err) {
+		log.Errorw(err, "error getting last census ID")
```
Comment (p4u) [inline api/censuses.go:171]: Don't print the error if you return it

File: api/censuses.go [hunk 5]
```diff
@@ -52,82 +57,139 @@ func (capi *census3API) getCensus(msg *api.APIdata, ctx *httprouter.HTTPContext)
...
 			log.Errorf("no strategy found for id %d: %s", req.StrategyID, err.Error())
-			return ErrNotFoundStrategy
+			return -1, ErrNoStrategyTokens.With("no strategy found")
 		}
 		log.Errorf("error getting strategy with id %d: %s", req.StrategyID, err.Error())
-		return ErrCantGetStrategy
+		return -1, ErrCantCreateCensus.With("error getting strategy")
+	}
+	if len(strategyTokens) == 0 {
+		log.Errorf("no tokens for strategy %d", req.StrategyID)
```
Comment (p4u) [inline api/censuses.go:163]: Don't print the error if you return it

File: api/censuses.go [hunk 6]
```diff
@@ -138,66 +200,135 @@ func (capi *census3API) createAndPublishCensus(msg *api.APIdata, ctx *httprouter
...
-		Size:       int64(len(strategyHolders)),
-		Weight:     censusWeight.Bytes(),
+		Size:       sqlCensusSize,
+		Weight:     sqlCensusWeight,
+		QueueID:    qID,
 	})
 	if err != nil {
-		log.Errorw(err, "error saving the census on the database")
-		return ErrCantCreateCensus
+		log.Errorf("error saving the census on the database: %v", err)
```
Comment (p4u) [inline api/censuses.go:251]: buh

File: api/censuses.go [hunk 7]
```diff
@@ -138,66 +200,135 @@ func (capi *census3API) createAndPublishCensus(msg *api.APIdata, ctx *httprouter
...
-	})
+	return newCensus.ID, nil
+}
+
+// getEnqueueCensus handler returns the current status of the queue item
+// identified by the ID provided. If it not exists it returns that the census
+// is not found. Else if the census exists and has been successfully created, it
+// will be included into the response. If not, the response only will include
+// if it is done or not and the resulting error.
+func (capi *census3API) getEnqueueCensus(msg *api.APIdata, ctx *httprouter.HTTPContext) error {
```
Comment (p4u) [inline api/censuses.go:266]: `s/"get"/""/g`

File: api/censuses.go [hunk 8]
```diff
@@ -138,66 +200,135 @@ func (capi *census3API) createAndPublishCensus(msg *api.APIdata, ctx *httprouter
...
+	if !exists {
+		return ErrNotFoundCensus.Withf("the ID %s does not exist in the queue", queueID)
+	}
+	// init queue item response
+	queueItem := QueueItemResponse{
+		Done:  done,
+		Error: err,
+	}
+	// check if it is not finished or some error occurred
+	if done && err == nil {
```
Comment (p4u) [inline api/censuses.go:282]: You are not checking this error.

File: api/censuses.go [hunk 9]
```diff
@@ -138,66 +200,135 @@ func (capi *census3API) createAndPublishCensus(msg *api.APIdata, ctx *httprouter
...
+		// if everything is ok, get the census information an return it
+		internalCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		censusID, ok := data["censusID"].(int)
+		if !ok {
+			log.Errorf("no census id registered on queue item: %v", err)
```
Comment (p4u) [inline api/censuses.go:288]: buh

File: api/types.go [hunk 10]
```diff
@@ -86,3 +88,9 @@ type GetStrategyResponse struct {
 	Tokens    []GetStrategyToken `json:"tokens"`
 	Predicate string             `json:"strategy"`
 }
+
+type QueueItemResponse struct {
+	Done   bool               `json:"done"`
+	Error  error              `json:"error"`
+	Census *GetCensusResponse `json:"census,omitempty"`
```
Comment (p4u) [inline api/types.go:95]: just `CensusResponse` or better something such as `CensusQueueResponse`.

Btw, does it actually need to be a pointer and to have omitempty? Try to use omitempty only when necessary.

File: queue/queue_test.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
+package queue
```
Comment (p4u) [inline queue/queue_test.go:1]: I'd like to have a test that runs all the Queue operations and with some mocked data if possible. 
Then we can run it with `-race` for some hours to be certain there are non data races.

File: api/censuses.go (hunk 3 above)
Comment (lucasmenendez) [reply api/censuses.go:191]: But probably the original error trace is useful for finding the root cause and at the same time, but does not provide any relevant information for the user of the API. This is why I usually print the error and return an "error response".

Comment (p4u) [reply api/censuses.go:191]: Why not returning the full error (with details) to the API? My concern is that we are treating twice, in a different manner, the same error.

**Review by p4u — APPROVED** [review]

**Pull request merged**